module/mempool/stdmap/backdata/arrayBackData.go

```diff
@@ -0,0 +1,400 @@
+package backdata
```

Comment:
`heropoolbykey` does not preserve Golang's [best practices on package naming](https://go.dev/blog/package-names?utm_source=dlvr.it&utm_medium=twitter), in specific, it is an unnecessarily long package name. Inspired by your comment, I re-organized the packaging of `heroCache` in  https://github.com/onflow/flow-go/pull/1778/commits/0ad7499fd52b2fe38a2c46ce0f73da8ae4fa756a and https://github.com/onflow/flow-go/pull/1778/commits/650b8f46ab99458556af67f48437676f270539d9 inline with the Golang best practices around package naming:

- We have `backdata/herocache` encapsulating the entire implementation of `heroCache`. 
- Renamed `backdata/arrayBackData.go` to `herocache/cache.go`, hence hero cache is available externally as `herocache.Cache`. 
- Renamed `backdata/heropool` to `herocache/pool` hence the underlying pool is accessible externally as `herocache.Pool`. 

With this new package naming we have a clear encapsulation of `heroCache` with two separate modules of `herocache.Cache` and `herocache.Pool`, which also follows the Golang package naming best practices.

Reply:
I think it's a step in the right direction but I still personally find it confusing and I will try to explain why:

I like the name 'pool' because it isn't obvious how you access the 'pool', i.e. it's not obvious that it's an index as with 'slice' or array, or a key as with 'map' or hash table. In fact, 'pool' could mean and was originally suggested as a name because it is accessible in this implementation via index or key... or other means indirectly, e.g. state :-)

However, the latest naming convention strips away that idea of generic 'pool' access via key or index because the access by key code is called `herocache/cache.go` and the access by index code is called `herocache/heropool/pool.go`.

And then to add further confusion, the 'cache' part of herocache suggests some kind of LRU since typically caches cache a subset of some larger set of data. However, I imagined the caching to be just another feature of heropool, and that feature is the ability to LRU or randomly eject an entity if the pool is full.

But the ejection feature should / could ideally be requested upon initialization, e.g. pseudo code `NewHeroPool(NumberOfStates, IncludeAccessByKey + IncludeCacheLRU)`. And similarly if the caller wants a heropool without key access and LRU then something like `NewHeroPool(NumberOfStates, ExcludeAccessByKey + ExcludeCacheLRU)` should / would also be possible. Here, trying to add an entity to a full pool would not work, nor would accessing the pool by key.

So if heropool supported access by key, state, or index then its API might look something like:

`addByKey(entityID, entity)` <-- equivalent of current cache `add()`
`addByState(entityId, entity, stateFree, stateUsed)` <-- equivalent of current pool `add()`
`addByIndex(entityId, entity, poolIndex)` <-- no equivalent yet...

And the current herocache would just be a thin wrapper initializing heropool with access by key and the LRU type, and mapping e.g. heropool `addByKey()` to the `backdata` `Add()` function, etc?

This way heropool does not have to be a cache / LRU but can optionally access pool entities by key, i.e. it is more abstract which means it's easier to use elsewhere?

Reply:
Thanks for the detailed comment. However, it looks like out of the scope of what this PR intends. This PR introduces `heroCache` as an alternative implementation of `BackData` interface, and the implementation correctly sticks to the defined interface. Extra abstractions of the `heroCache` that this comment suggests are more suitable to the state-based version of the hero cache, which has been planned as the next steps. Hence, adding your comment to https://github.com/dapperlabs/flow-go/issues/6083 to be considered later.